flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -1872,9 +2054,34 @@ class ScalarExprLowering {
         continue;
       }
       // Ad-hoc argument lowering handling.
-      auto lowerAs = Fortran::lower::lowerIntrinsicArgumentAs(
-          getLoc(), *argLowering, dummy.name);
-      switch (lowerAs) {
+      mlir::Location loc = getLoc();
+      Fortran::lower::ArgLoweringRule argRules =
+          Fortran::lower::lowerIntrinsicArgumentAs(loc, *argLowering,
+                                                   dummy.name);
+      if (argRules.handleDynamicOptional &&
+          mayBePassedAsAbsentOptional(*expr, converter.getFoldingContext())) {
+        ExtValue optional = lowerIntrinsicArgumentAsInquired(*expr);
+        mlir::Value isPresent = genActualIsPresentTest(builder, loc, optional);
+        switch (argRules.lowerAs) {
+        case Fortran::lower::LowerIntrinsicArgAs::Value:
+          operands.emplace_back(
+              genOptionalValue(builder, loc, optional, isPresent));
+          continue;
```

Comment:
The 3 cases are textually identical, I believe. Rather than block copying the code, you can just do
```
switch (...) {
  case a:
  case b:
  case c:
    <common code here>
    continue;
```